Source/Core/Core/IOS/ES/ES.cpp

```diff
@@ -613,37 +628,91 @@ IPCCommandResult ES::AddContentFinish(const IOCtlVRequest& request)
 
-  std::string content_path;
-  if (content_info.IsShared())
-  {
-    DiscIO::CSharedContent shared_content{Common::FROM_SESSION_ROOT};
-    content_path = shared_content.AddSharedContent(content_info.sha1.data());
-  }
-  else
+  // Just write all contents to the title import directory. AddTitleFinish will
+  // move the contents to the proper location.
+  const std::string tmp_path = GetImportContentPath(m_addtitle_tmd, m_addtitle_content_id);
+  File::CreateFullPath(tmp_path);
+
+  File::IOFile fp(tmp_path, "wb");
+  if (!fp.WriteBytes(decrypted_data.data(), content_info.size))
   {
-    content_path = StringFromFormat(
-        "%s%08x.app",
-        Common::GetTitleContentPath(m_addtitle_tmd.GetTitleId(), Common::FROM_SESSION_ROOT).c_str(),
-        m_addtitle_content_id);
+    ERROR_LOG(IOS_ES, "AddContentFinish: Failed to write to %s", tmp_path.c_str());
+    return GetDefaultReply(ES_WRITE_FAILURE);
   }
 
-  File::IOFile fp(content_path, "wb");
-  fp.WriteBytes(decrypted_data.data(), content_info.size);
-
   m_addtitle_content_id = 0xFFFFFFFF;
   return GetDefaultReply(IPC_SUCCESS);
 }
 
+static void AbortImport(const u64 title_id, const std::vector<std::string>& processed_paths)
+{
+  for (const auto& path : processed_paths)
+    File::Delete(path);
+
+  const std::string import_dir = Common::GetImportTitlePath(title_id);
+  File::DeleteDirRecursively(import_dir);
+}
+
 IPCCommandResult ES::AddTitleFinish(const IOCtlVRequest& request)
 {
   if (!request.HasNumberOfValidVectors(0, 0) || !m_addtitle_tmd.IsValid())
     return GetDefaultReply(ES_PARAMETER_SIZE_OR_ALIGNMENT);
 
+  std::vector<std::string> processed_paths;
+
+  for (const auto& content_info : m_addtitle_tmd.GetContents())
+  {
+    const std::string source = GetImportContentPath(m_addtitle_tmd, content_info.id);
+
+    // Contents may not have been all imported. This is normal and this isn't an error condition.
+    if (!File::Exists(source))
+      continue;
+
+    std::string content_path;
+    if (content_info.IsShared())
+    {
+      DiscIO::CSharedContent shared_content{Common::FROM_SESSION_ROOT};
+      content_path = shared_content.AddSharedContent(content_info.sha1.data());
+    }
+    else
+    {
+      content_path =
+          StringFromFormat("%s%08x.app", Common::GetTitleContentPath(m_addtitle_tmd.GetTitleId(),
+                                                                     Common::FROM_SESSION_ROOT)
+                                             .c_str(),
+                           content_info.id);
+    }
+
+    File::CreateFullPath(content_path);
+    if (!File::RenameSync(source, content_path))
+    {
+      ERROR_LOG(IOS_ES, "AddTitleFinish: Failed to rename %s to %s", source.c_str(),
+                content_path.c_str());
+      AbortImport(m_addtitle_tmd.GetTitleId(), processed_paths);
+      return GetDefaultReply(ES_WRITE_FAILURE);
+    }
+
+    processed_paths.push_back(content_path);
```

Comment:
Not sure, we'd need to remove the entry from the shared content map as well. The reason we're removing private contents is that they can cause issues if a title is incomplete; we don't have that issue with shared contents.